src/main/java/org/kohsuke/groovy/sandbox/impl/Checker.java

```diff
@@ -746,4 +957,56 @@ private static boolean isCollectionSafeToCast(Collection c) {
+     *
+     * <p>Groovy uses runtime APIs (e.g. {@link ScriptByteCodeAdapter}) to support various standard language features
+     * such as unary operators. Some of these APIs invoke methods reflectively based on the runtime types of arguments,
+     * which the sandbox does not see and so it cannot intercept those calls. We define sandbox-aware replacements for
+     * these methods so that these reflective calls can be intercepted.
+     * <p>When using groovy-sandbox through script-security without groovy-cps, these replacements only take effect if
+     * a script directly calls one of the original methods. When using groovy-cps, these replacements also take effect
+     * when their corresponding AST nodes (e.g. unary operators) are used.
+     */
+    private static Thunk findCheckedReplacement(Class<?> clazz, String method, Object[] args) {
+        Method maybeReplacement = GROOVY_RUNTIME_REPLACEMENTS.get(new SimpleImmutableEntry(clazz, method));
+        if (maybeReplacement == null) {
+            return null;
+        }
+        ParameterTypes parameterTypes = new ParameterTypes(maybeReplacement.getParameterTypes());
+        if (!parameterTypes.isValidExactMethod(args)) {
+            return null;
+        }
+        return () -> {
+            try {
+                return maybeReplacement.invoke(null, args);
+            } catch (InvocationTargetException e) {
+                throw e.getCause(); // e.g. CpsCallableInvocation
+            }
+        };
+    }
+
+    /**
+     * A map from Groovy methods to checked replacements that will be used instead when the sandbox is active.
+     */
+    private static final HashMap<Map.Entry<Class<?>, String>, Method> GROOVY_RUNTIME_REPLACEMENTS = new HashMap<>();
+    static {
+        addReplacement(InvokerHelper.class, "bitwiseNegate", "checkedBitwiseNegate", Object.class);
+        addReplacement(InvokerHelper.class, "unaryMinus", "checkedUnaryMinus", Object.class);
+        addReplacement(InvokerHelper.class, "unaryPlus", "checkedUnaryPlus", Object.class);
+        addReplacement(ScriptBytecodeAdapter.class, "bitwiseNegate", "checkedBitwiseNegate", Object.class);
+        addReplacement(ScriptBytecodeAdapter.class, "unaryMinus", "checkedUnaryMinus", Object.class);
+        addReplacement(ScriptBytecodeAdapter.class, "unaryPlus", "checkedUnaryPlus", Object.class);
+        addReplacement(ScriptBytecodeAdapter.class, "createRange", "checkedCreateRange", Object.class, Object.class, boolean.class);
+    }
+
+    private static void addReplacement(Class<?> clazz, String name, String checkedName, Class<?>... parameterTypes) {
+        try {
+            GROOVY_RUNTIME_REPLACEMENTS.put(new SimpleImmutableEntry(clazz, name), Checker.class.getDeclaredMethod(checkedName, parameterTypes));
+        } catch (NoSuchMethodException e) {
+            throw new AssertionError(e); // Developer error.
```

Comment:
Rather than keeping `Method`s, it would be more type-safe (no reflection needed) to define a SAM: `Function<Object[], Object>`. E.g.

```java
args -> checkedCreateRange(args[0], args[1], (Boolean) args[2])
```